x/wasm/internal/keeper/keeper.go

```diff
@@ -469,7 +473,48 @@ func (k Keeper) Sudo(ctx sdk.Context, contractAddress sdk.AccAddress, msg []byte
 	events := types.ParseEvents(res.Attributes, contractAddress)
 	ctx.EventManager().EmitEvents(events)
 
-	err = k.dispatchMessages(ctx, contractAddress, contractInfo.IBCPortID, res.Messages)
+	// dispatch submessages then messages
+	err = k.dispatchAll(ctx, contractAddress, contractInfo.IBCPortID, res.Submessages, res.Messages)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "dispatch")
+	}
+
+	return &sdk.Result{
+		Data: res.Data,
+	}, nil
+}
+
+// reply is only called from keeper internal functions (dispatchSubmessages) after processing the submessage
+// it
+func (k Keeper) reply(ctx sdk.Context, contractAddress sdk.AccAddress, reply wasmvmtypes.Reply) (*sdk.Result, error) {
+	// we can assume this is still in cache and not charge again for returning the reply. correct?
```

Comment:
The LRU cache might be disabled via `--wasm.memory_cache_size` start parameter.
I have mixed feelings not charging anything or charging the full instantiation fee (again). Let's discuss in a bigger group.

Reply:
Decision was to charge the same fee here (which is 0 if pinned).
We can consider some discount later